datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java

```diff
@@ -1046,6 +1059,130 @@ public int cleanUpOrphanConnectorTasks(boolean cleanUpOrphanTasksInConnector) {
     return orphanCount;
   }
 
+  private Set<DatastreamTask> getDatastreamTasks() {
+    return getAllAssignedDatastreamTasks().values().stream().flatMap(Collection::stream).collect(Collectors.toSet());
+  }
+
+  /**
+   * Identify orphan connector task locks for which there is no corresponding connector task node present and schedule
+   * the clean up after the debounce timer. There is no way for the leader to know that the task corresponding to the
+   * lock has stopped or not. It makes sense to clear it after the debounce timer.
+   *
+   * NOTE: this should be called after the valid tasks have been reassigned or become safe to discard per
+   * strategy requirement.
+   *
+   * This is a costly operation which involves getting all children of /cluster/connectors/lock from Zookeeper. So,
+   * it should be called only once the leader get elected and has finished the assignment and cleaned up dead tasks.
+   * @param cleanUpOrphanTaskLocksInConnector Boolean whether orphan task locks should be removed from zookeeper or just
+   *                                      print warning logs.
+   * @return total orphan task locks identified/cleaned.
+   */
+  public int cleanUpOrphanConnectorTaskLocks(boolean cleanUpOrphanTaskLocksInConnector) {
+    // do not perform this operation if not a leader or another operation is going on.
+    if (!_isLeader || !_orphanLockCleanupFuture.isDone()) {
+      return 0;
+    }
+
+    Set<DatastreamTask> validTasks = getDatastreamTasks();
+    List<String> allConnectors = getAllConnectors();
+
+    int orphanCount = 0;
+    for (String connector : allConnectors) {
+      Map<String, Set<String>> locksByTaskPrefix = getAllConnectorTaskLocks(connector);
+
+      Set<String> validTaskNamesSet = validTasks.stream()
+          .filter(x -> x.getConnectorType().equals(connector))
+          .map(DatastreamTask::getDatastreamTaskName)
+          .collect(Collectors.toSet());
+
+      List<String> orphanLockList = new ArrayList<>();
+
+      locksByTaskPrefix.forEach((taskPrefix, taskList) -> {
+        // if all the task locks in the task prefix needs to be cleaned up, delete the task prefix node as well.
+        AtomicBoolean deleteTaskPrefixNode = new AtomicBoolean(true);
+        taskList.forEach(taskName -> {
+          if (!validTaskNamesSet.contains(taskName)) {
+            orphanLockList.add(KeyBuilder.datastreamTaskLock(_cluster, connector, taskPrefix, taskName));
+          } else {
+            deleteTaskPrefixNode.set(false);
+          }
+        });
+
+        if (deleteTaskPrefixNode.get()) {
+          orphanLockList.add(KeyBuilder.datastreamTaskLockPrefix(_cluster, connector, taskPrefix));
+        }
```

Comment:
I initially thought of doing this but decided not to do it, since I did not want to flatten the task list and keep it in the similar structure. Also, we need taskPrefix which we will not get from the String.

Reply:
Sure